File: app/src/main/java/com/battlelancer/seriesguide/traktapi/GenericCheckInDialogFragment.java
package com.battlelancer.seriesguide.traktapi;

import android.content.DialogInterface;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDialogFragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.ui.SeriesGuidePreferences;
import com.battlelancer.seriesguide.util.Utils;
import com.google.android.material.textfield.TextInputLayout;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

public abstract class GenericCheckInDialogFragment extends AppCompatDialogFragment {

    public interface InitBundle {

        /**
         * Title of episode or movie. <b>Required.</b>
         */
        String ITEM_TITLE = "itemtitle";

        /**
         * Movie TMDb id. <b>Required for movies.</b>
         */
        String MOVIE_TMDB_ID = "movietmdbid";

        /**
         * Season number. <b>Required for episodes.</b>
         */
        String EPISODE_TVDB_ID = "episodetvdbid";
    }

    public class CheckInDialogDismissedEvent {
    }

    @BindView(R.id.textInputLayoutCheckIn) TextInputLayout textInputLayout;
    @BindView(R.id.buttonCheckIn) View buttonCheckIn;
    @BindView(R.id.buttonCheckInPasteTitle) View buttonPasteTitle;
    @BindView(R.id.buttonCheckInClear) View buttonClear;
    @BindView(R.id.progressBarCheckIn) View progressBar;

    private Unbinder unbinder;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // hide title, use special theme with exit animation
        setStyle(STYLE_NO_TITLE, R.style.Theme_SeriesGuide_Dialog_CheckIn);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.dialog_checkin, container, false);
        unbinder = ButterKnife.bind(this, view);

        // Paste episode button
        final String itemTitle = getArguments().getString(InitBundle.ITEM_TITLE);
        final EditText editTextMessage = textInputLayout.getEditText();
        if (!TextUtils.isEmpty(itemTitle)) {
            buttonPasteTitle.setOnClickListener(v -> {
                if (editTextMessage == null) {
                    return;
                }
                int start = editTextMessage.getSelectionStart();
                int end = editTextMessage.getSelectionEnd();
                editTextMessage.getText().replace(Math.min(start, end), Math.max(start, end),
                        itemTitle, 0, itemTitle.length());
            });
        }

        // Clear button
        buttonClear.setOnClickListener(v -> {
            if (editTextMessage == null) {
                return;
            }
            editTextMessage.setText(null);
        });

        // Checkin Button
        buttonCheckIn.setOnClickListener(v -> checkIn());

        setProgressLock(false);

        return view;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        // immediately start to check-in if the user has opted to skip entering a check-in message
        if (TraktSettings.useQuickCheckin(getContext())) {
            checkIn();
        }
    }

    @Override
    public void onStart() {
        super.onStart();

        EventBus.getDefault().register(this);
    }

    @Override
    public void onStop() {
        super.onStop();

        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);

        EventBus.getDefault().post(new CheckInDialogDismissedEvent());
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();

        unbinder.unbind();
    }

    @SuppressWarnings("unused")
    @Subscribe
    public void onEvent(TraktTask.TraktActionCompleteEvent event) {
        // done with checking in, unlock UI
        setProgressLock(false);

        if (event.wasSuccessful) {
            // all went well, dismiss ourselves
            dismissAllowingStateLoss();
        }
    }

    @SuppressWarnings("unused")
    @Subscribe
    public void onEvent(TraktTask.TraktCheckInBlockedEvent event) {
        // launch a check-in override dialog
        TraktCancelCheckinDialogFragment
                .show(getFragmentManager(), event.traktTaskArgs, event.waitMinutes);
    }

    private void checkIn() {
        // lock down UI
        setProgressLock(true);

        // connected?
        if (Utils.isNotConnected(getActivity())) {
            // no? abort
            setProgressLock(false);
            return;
        }

        // launch connect flow if trakt is not connected
        if (!TraktCredentials.ensureCredentials(getActivity())) {
            // not connected? abort
            setProgressLock(false);
            return;
        }

        // try to check in
        EditText editText = textInputLayout.getEditText();
        if (editText != null) {
            checkInTrakt(editText.getText().toString());
        }
    }

    /**
     * Start the trakt check-in task.
     */
    protected abstract void checkInTrakt(String message);

    /**
     * Disables all interactive UI elements and shows a progress indicator.
     */
    private void setProgressLock(boolean lock) {
        progressBar.setVisibility(lock ? View.VISIBLE : View.GONE);
        textInputLayout.setEnabled(!lock);
        buttonPasteTitle.setEnabled(!lock);
        buttonClear.setEnabled(!lock);
        buttonCheckIn.setEnabled(!lock);
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/traktapi/GenericCheckInDialogFragment.kt
package com.battlelancer.seriesguide.traktapi

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.lifecycle.lifecycleScope
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.DialogCheckinBinding
import com.battlelancer.seriesguide.traktapi.TraktTask.TraktActionCompleteEvent
import com.battlelancer.seriesguide.traktapi.TraktTask.TraktCheckInBlockedEvent
import com.battlelancer.seriesguide.util.Utils
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import kotlin.math.max
import kotlin.math.min

abstract class GenericCheckInDialogFragment : AppCompatDialogFragment() {

    class CheckInDialogDismissedEvent

    private var binding: DialogCheckinBinding? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val binding = DialogCheckinBinding.inflate(layoutInflater)
        this.binding = binding

        // Paste episode button
        val itemTitle = requireArguments().getString(ARG_ITEM_TITLE)
        val editTextMessage = binding.textInputLayoutCheckIn.editText
        if (!itemTitle.isNullOrEmpty()) {
            binding.buttonCheckInPasteTitle.setOnClickListener {
                if (editTextMessage == null) {
                    return@setOnClickListener
                }
                val start = editTextMessage.selectionStart
                val end = editTextMessage.selectionEnd
                editTextMessage.text.replace(
                    min(start, end), max(start, end),
                    itemTitle, 0, itemTitle.length
                )
            }
        }

        // Clear button
        binding.buttonCheckInClear.setOnClickListener {
            if (editTextMessage == null) {
                return@setOnClickListener
            }
            editTextMessage.text = null
        }

        // Checkin Button
        binding.buttonCheckIn.setOnClickListener { checkIn() }

        setProgressLock(false)

        lifecycleScope.launchWhenStarted {
            // immediately start to check-in if the user has opted to skip entering a check-in message
            if (TraktSettings.useQuickCheckin(requireContext())) {
                checkIn()
            }
        }

        return MaterialAlertDialogBuilder(requireContext(), R.style.Theme_SeriesGuide_Dialog_CheckIn)
            .setView(binding.root)
            .create()
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        EventBus.getDefault().post(CheckInDialogDismissedEvent())
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    @Subscribe
    fun onEvent(event: TraktActionCompleteEvent) {
        // done with checking in, unlock UI
        setProgressLock(false)
        if (event.wasSuccessful) {
            // all went well, dismiss ourselves
            dismissAllowingStateLoss()
        }
    }

    @Subscribe
    fun onEvent(event: TraktCheckInBlockedEvent) {
        // launch a check-in override dialog
        TraktCancelCheckinDialogFragment
            .show(parentFragmentManager, event.traktTaskArgs, event.waitMinutes)
    }

    private fun checkIn() {
        // lock down UI
        setProgressLock(true)

        // connected?
        if (Utils.isNotConnected(requireContext())) {
            // no? abort
            setProgressLock(false)
            return
        }

        // launch connect flow if trakt is not connected
        if (!TraktCredentials.ensureCredentials(requireContext())) {
            // not connected? abort
            setProgressLock(false)
            return
        }

        // try to check in
        val editText = binding?.textInputLayoutCheckIn?.editText
        if (editText != null) {
            checkInTrakt(editText.text.toString())
        }
    }

    /**
     * Start the Trakt check-in task.
     */
    protected abstract fun checkInTrakt(message: String)

    /**
     * Disables all interactive UI elements and shows a progress indicator.
     */
    private fun setProgressLock(lock: Boolean) {
        val binding = binding ?: return
        binding.progressBarCheckIn.visibility = if (lock) View.VISIBLE else View.GONE
        binding.textInputLayoutCheckIn.isEnabled = !lock
        binding.buttonCheckInPasteTitle.isEnabled = !lock
        binding.buttonCheckInClear.isEnabled = !lock
        binding.buttonCheckIn.isEnabled = !lock
    }

    companion object {
        /**
         * Title of episode or movie. **Required.**
         */
        const val ARG_ITEM_TITLE = "itemtitle"

        /**
         * Movie TMDb id. **Required for movies.**
         */
        const val ARG_MOVIE_TMDB_ID = "movietmdbid"
        const val ARG_EPISODE_ID = "episodeid"
    }
}